File: sr_demo/gen_window.m
function w=gen_window(s,d1,d2);
%size s is an array s(1) = x, s(2) = y

if ~exist('d1','var')
    d1=0.1;
end
if ~exist('d2','var');
    d2=0.1;
end

% Create a s(1) by 1 matrix (column matrix) filled with ones
x=ones(s(1),1);
% Create a linear vector (row matrix) with s(1)*d1 evenly spaced numbers
%between 0 and 1
slope=linspace(0,1,round(s(1)*d1));
% Replace values in x with values in slope for full length of slope
x(1:length(slope))=slope;
% flip x around middle of x (assuming small d1 relative to s(1))
% ex: a,b,c,1,1,1,1,1,1,1,1,1,1
%    results in array a,b,c,1,1,1,1,1,1,1,c,b,a
x(length(x)-length(slope)+1:length(x))=fliplr(slope);

% Same for y
y=ones(1,s(2));
slope=linspace(0,1,round(s(2)*d2));
y(1:length(slope))=slope;
y(length(y)-length(slope)+1:length(y))=fliplr(slope);

% Results in a s(1) by s(2) matrix where all borders are 0 and the middle
%is 1 with s(1)*d1 numbers fading from 1 to 0 at each x edge and s(2)*d2 at
%each y edge
w=x*y;

File: src/final/gen_window.cpp
#include "gen_window.h"
#include <vector>

using namespace std;
using namespace cv;

/******************************************************************************

	gen_window generates a matrix of size [height x width] with [channels] 
	channels.  The generated matrix fades linearally from 0 (black) to 1 
	(white) across hpercent percent of the top and bottom borders and wpercent
	percent of the left and right borders.  This is creates a kernel that we
	can use to detect edges without running into problems with the edges of the
	matrix itself.


	gen_window:
		int height:
			The height of the desired output Mat.
		int width:
			The width of the desired output Mat.
		float hpercent:
			The percent of the top and bottom images to fade to black on those 
			edges of the image.
		float wpercent:
			The percent of the left and right images to fade to black on those 
			edges of the image.

******************************************************************************/
Mat gen_window( int height, 
				int width, 
				float hpercent, 
				float wpercent, 
				int channels ){
	vector<Mat> planes;
	Mat retVal;

	//Create an height by 1 matrix (column matrix) filled with ones
	Mat x = Mat::ones(width, 1, CV_32F);

	int steps = width * wpercent;
	float slope = 0;
	float increment = 1.0/(steps - 1);

	/*	Create a linear vector (row matrix) with width*wpercent evenly spaced 
	numbers	between 0 and 1; a, b, c, etc.
	results in array a,b,c,1,1,1,1,1,1,1,c,b,a */
	for(int i=0; i<steps; i++){
		x.at<float>(i, 0) = slope;
		x.at<float>(width - 1 - i, 0) = slope;
		slope = slope + increment;
	}

	Mat y = Mat::ones(1, height, CV_32F);
	steps = height * hpercent;
	slope = 0;
	increment = 1.0/(steps - 1);

	/*	Create a linear vector (row matrix) with height*hpercent evenly spaced 
	numbers between 0 and 1; a, b, c, etc.
	results in array a,b,c,1,1,1,1,1,1,1,c,b,a */
	for(int i=0; i<steps; i++){
		y.at<float>(0, i) = slope;
		y.at<float>(0, height - 1 - i) = slope;
		slope = slope + increment;
	}

	Mat z = Mat(width, height, CV_32F);
	z = x * y;

	for(int i = 0; i < channels; i++)
		planes.push_back(z);

	merge(planes, retVal);

	return retVal;
}